expression/builtin_time.go

```diff
@@ -2078,6 +2122,42 @@ func (b *builtinTimestamp2ArgsSig) evalTime(row []types.Datum) (types.Time, bool
 	return result, false, nil
 }
 
+type timestampLiteralFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *timestampLiteralFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpDatetime, tpString)
+	bf.tp.Tp, bf.tp.Flen, bf.tp.Decimal = mysql.TypeTimestamp, 19, 0
+	sig := &builtinTimestampLiteralSig{baseTimeBuiltinFunc{bf}}
+	return sig.setSelf(sig), nil
+}
+
+type builtinTimestampLiteralSig struct {
+	baseTimeBuiltinFunc
+}
+
+// evalTime evals TIMESTAMP 'stringLit'.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-literals.html
+func (b *builtinTimestampLiteralSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	str, isNull, err := b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return types.Time{}, true, errors.Trace(err)
+	}
+	if !TimestampPattern.MatchString(str) {
```

Comment:
can we skip this? just use types.ParseTime to check whether the literral is illegal?

Reply:
I think this can't be skipped. `TIME`, `TIMESTAMP`, `DATE` literal have more strict format check than other conditions. for example, `2017-01-01 00:00:00` can be cast to type `DATE`, but use `2017-01-01 00:00:00` in `DATE` literal is illegal. The same as `2017-01-01 00:00:00` on `TIME` literal.
```sql
mysql> select date '2017-01-01 00:00:00';
ERROR 1525 (HY000): Incorrect DATE value: '2017-01-01 00:00:00'
mysql> select cast('2017-01-01 00:00:00' as date);                                                                                             
 
Field   1:  `cast('2017-01-01 00:00:00' as date)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATE
Collation:  binary (63)
Length:     10
Max_length: 10
Decimals:   0
Flags:      BINARY 


+-------------------------------------+
| cast('2017-01-01 00:00:00' as date) |
+-------------------------------------+
| 2017-01-01                          |
+-------------------------------------+
1 row in set (0.00 sec)

```

```sql
mysql> select time '2017-01-01 00:00:00';
ERROR 1525 (HY000): Incorrect TIME value: '2017-01-01 00:00:00'
mysql> select cast('2017-01-01 00:00:00' as time);
Field   1:  `cast('2017-01-01 00:00:00' as time)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       TIME
Collation:  binary (63)
Length:     10
Max_length: 8
Decimals:   0
Flags:      BINARY 


+-------------------------------------+
| cast('2017-01-01 00:00:00' as time) |
+-------------------------------------+
| 00:00:00                            |
+-------------------------------------+
1 row in set (0.00 sec)

```